rubix-spi/src/main/java/com/qubole/rubix/spi/fop/ObjectPool.java

```diff
@@ -0,0 +1,132 @@
+ * limitations under the License. See accompanying LICENSE file.
+ * <p>
+ * <p>
+ * NOTICE: THIS FILE HAS BEEN MODIFIED BY  Qubole Inc UNDER COMPLIANCE WITH THE APACHE 2.0 LICENCE FROM THE ORIGINAL WORK
+ * OF https://github.com/DanielYWoo/fast-object-pool.
+ */
+package com.qubole.rubix.spi.fop;
+
+import com.qubole.rubix.spi.CacheConfig;
+import com.qubole.rubix.spi.RetryingPooledBookkeeperClient;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.conf.Configuration;
+
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * @author Daniel
+ */
+public class ObjectPool<T>
+{
+  private static final Log log = LogFactory.getLog(RetryingPooledBookkeeperClient.class);
+
+  private final PoolConfig config;
+  private final ObjectFactory<T> factory;
+  private final ConcurrentHashMap<String, ObjectPoolPartition<T>> hostToPoolMap;
+
+  public ObjectPool(PoolConfig poolConfig, ObjectFactory<T> objectFactory)
+  {
+    this.config = poolConfig;
+    this.factory = objectFactory;
+    this.hostToPoolMap = new ConcurrentHashMap<>();
+  }
+
+  public void registerHost(String host, int socketTimeout, int connectTimeout)
+  {
+    hostToPoolMap.put(host, new ObjectPoolPartition<>(this, config, factory, createBlockingQueue(config), host, socketTimeout, connectTimeout));
+  }
+
+  protected BlockingQueue<Poolable<T>> createBlockingQueue(PoolConfig poolConfig)
+  {
+    return new ArrayBlockingQueue<>(poolConfig.getMaxSize());
+  }
+
+  public Poolable<T> borrowObject(String host, Configuration conf)
+  {
+    if (!hostToPoolMap.containsKey(host)) {
+      synchronized (hostToPoolMap) {
+        if (!hostToPoolMap.containsKey(host)) {
+          int socketTimeout = CacheConfig.getServerSocketTimeout(conf);
+          int connectTimeout = CacheConfig.getServerConnectTimeout(conf);
+          registerHost(host, socketTimeout, connectTimeout);
+        }
+      }
+    }
+    log.debug("Borrowing object for partition: " + host);
+    for (int i = 0; i < 3; i++) { // try at most three times
+      Poolable<T> result = getObject(true, host);
+      if (factory.validate(result.getObject())) {
+        return result;
+      }
+      else {
+        this.hostToPoolMap.get(host).decreaseObject(result);
+      }
+    }
+    throw new RuntimeException("Cannot find a valid object");
+  }
+
+  private Poolable<T> getObject(boolean blocking, String host)
+  {
+    ObjectPoolPartition<T> subPool = this.hostToPoolMap.get(host);
+
+    Poolable<T> freeObject;
+    if (subPool.getObjectQueue().size() == 0) {
+      // increase objects and return one, it will return null if reach max size
+      subPool.increaseObjects(this.config.getDelta());
+    }
+
+    try {
+      if (blocking) {
+        log.warn("Pool is empty...Waiting for connection");
```

Comment:
Given this will be used in every call, warn level if too high. Infact, remote this log as take() doesnt necessarily mean waiting

Reply:
yeah, this is an old log, we should remove this.